### PR TITLE
feat: add devcontainer configuration

### LINKED ISSUE #2: Add tools and replace Jest with Vitest
There are some useful tools like:

- [DevContainer] and [Developing inside a Container]
- [Vitest]
- [Typedoc]
- [ESLint] with [opinionated config]
- stricter type checking with [ts-reset] and [centralized TSConfig]
- GitHub action for checking code before publishing code onto `npm`
- update `package.json` properties

[Typedoc]: https://typedoc.org/
[Vitest]: https://vitest.dev/
[DevContainer]: https://containers.dev/
[Developing inside a Container]: https://code.visualstudio.com/docs/devcontainers/containers
[ESLint]: https://eslint.org/
[opinionated config]: https://github.com/antfu/eslint-config
[ts-reset]: https://github.com/total-typescript/ts-reset
[centralized TSConfig]: https://github.com/tsconfig/bases/tree/main

### PR DESCRIPTION
**Pull request by nandordudas**
I have separated the changes into small pieces the easier-to-understand.

What do you have?

- you can use new GitHub workflow items and you can restrict your pull request (under repository setting at branch rules)
- you have a release helper, so you only need to tag you code locally and push it to the remote, workflow will create a new release under your repository; its' recommended to add new tag rule under repository settings; **you need to follow conventional commiting rules**
- more strict static analyzer settings for Typescript and ESLint, you can modify it anytime
- documentation is generated by a 3rd party tool like `typedoc` instead of writing into README directly; you can add examples on method with `@example` annotation
- test are covered by `Vitest` and coverage is fully passed
- most of `for loops` are replaced by `forEach`, it's not necessary you can decide what you want
- callbacks are called without `.apply`
- explicit types are only used on sensitive parts
- Devcontainer configuration is available for VS Code

### CODE REVIEW

File: src/apt-map.ts [hunk 1]
```diff
@@ -1,218 +1,297 @@
...
+    }
+
+    return result
   }
 
-  concat<T extends V>(aptMap: AptMap<K, T>): AptMap<K, V> {
+  /**
+   * Creates a shallow copy of an AptMap and returns it.
+   */
+  clone() {
```
Comment (jarcko) [inline src/apt-map.ts]: All methods were sorted alphabetically, please keep the initial order. Thanks

File: src/apt-map.ts [hunk 2]
```diff
@@ -1,218 +1,297 @@
...
+   */
+  entriesAsArray() {
+    return Array.from(this.entries())
   }
 
-  every(cb: (value: V, key?: K, aptMap?: AptMap<K, V>) => boolean): boolean {
-    let result = false;
+  /**
+   * Filters input apt map using callback function and returns filtered apt map.
+   */
```
Comment (jarcko) [inline src/apt-map.ts:85]: All these typedocs seems to be redundant. Or at least they should be completed with `@param` `@returns` signatures

File: src/apt-map.ts [hunk 3]
```diff
@@ -1,218 +1,297 @@
...
+  filter(cb: Callback<K, V>) {
+    const filtered = new AptMap<K, V>()
 
-    for (const [mapKey, mapValue] of this) {
-      result = cb.apply(this, [mapValue, mapKey, this]);
+    this.forEach((mapValue, mapKey) => {
+      const isFiltered = cb(mapValue, mapKey, this)
```
Comment (jarcko) [inline src/apt-map.ts]: Please revert this to back `apply()` here and everywhere. It was added for a reason, we used to have an issue with loosing a context when first time written it this way.

File: src/apt-map.ts [hunk 4]
```diff
@@ -1,218 +1,297 @@
...
-    return result;
+    return filtered
   }
 
-  filter(cb: (value: V, key?: K, aptMap?: AptMap<K, V>) => boolean): AptMap<K, V> {
-    const filtered = new AptMap<K, V>();
+  /**
+   * Returns first entry which satisfies condition in callback function, otherwise an empty array.
+   */
+  findEntry(cb: Callback<K, V>) {
```
Comment (jarcko) [inline src/apt-map.ts]: Why all the methods lost return type?

File: src/apt-map.ts [hunk 5]
```diff
@@ -1,218 +1,297 @@
...
+  toObject() {
+    return this.remapToObject(value => value)
   }
 
-  valuesAsArray(): V[] {
-    return Array.from(this.values());
+  /**
+   * Returns an array of values.
+   */
+  valuesAsArray() {
```
Comment (jarcko) [inline src/apt-map.ts]: why all them methods and getters lost return type

File: src/apt-map.ts [hunk 6]
```diff
@@ -1,218 +1,297 @@
...
+      const isFound = cb(mapValue, mapKey, this)
 
-      if (isFiltered) {
-        filtered.set(mapKey, mapValue);
-      }
-    }
-
-    return filtered;
-  }
+      if (!isFound)
```
Comment (jarcko) [inline src/apt-map.ts]: Please always use curly brackets withing `if` statement

File: src/apt-map.ts [hunk 7]
```diff
@@ -1,218 +1,297 @@
+type Callback<K, V, R = boolean> = (value: V, key?: K, aptMap?: AptMap<K, V>) => R
+
 export class AptMap<K = any, V = any> extends Map<K, V> {
+  /**
+   * Returns first key if it exists, otherwise `undefined`.
+   */
   get firstKey(): K | undefined {
-    return this.keys().next().value;
+    return this.keys().next().value
```
Comment (jarcko) [inline src/apt-map.ts]: semicolon on each row is good practice

File: src/apt-map.ts [hunk 8]
```diff
@@ -1,218 +1,297 @@
...
+  /**
+   * Returns first entry which satisfies condition in callback function, otherwise an empty array.
+   */
+  findEntry(cb: Callback<K, V>) {
+    let foundKey: K | undefined
+    let foundValue: V | undefined
 
-    for (const [mapKey, mapValue] of this) {
-      const isFiltered = cb.apply(this, [mapValue, mapKey, this]);
+    this.forEach((mapValue, mapKey) => {
```
Comment (jarcko) [inline src/apt-map.ts]: With regular `for...of` we used `break;` which used to stop the loop when the item is found. Now, with your change it continues loop. So when you have 1000 entries and found entry is first entry you would still interate through other 999. 

That's why `for...of` was applied for all methods (even without `brake;`) just to have all in same style

**Review by jarcko — CHANGES_REQUESTED** [review]

File: src/apt-map.ts (hunk 4 above)
Comment (nandordudas) [reply src/apt-map.ts]: Types are inferred by language server, if you want to extend the class explicit return types are recommended.

File: src/apt-map.ts (hunk 7 above)
Comment (nandordudas) [reply src/apt-map.ts]: I'll update ESLint for these settings.

File: src/apt-map.ts (hunk 8 above)
Comment (nandordudas) [reply src/apt-map.ts]: image

You can check differences if you want.

File: src/apt-map.ts (hunk 2 above)
Comment (nandordudas) [reply src/apt-map.ts:85]: I just copy from the README file, I think most of comment are redundant but this not why I'm here.

File: src/apt-map.ts (hunk 5 above)
Comment (nandordudas) [reply src/apt-map.ts]: This is type inferencing, hover over the function and see the unknown magic behind the scenes.

image

**Comment by nandordudas** [comment]
You don't need to merge this request and you can edit too if you want. It contains tools which can be useful in the future for you, or anyone else.

In general, in relation to your code, I would have used more abstraction, especially in the case of the `toArray` and `toObject` methods. Also, these abstract classes can be used by others according to their requirements (don't forget to use the `override` keyword).

**Comment by jarcko** [comment]
@nandordudas Okay, thanks, as long as now it's a branch in this repo, I can adjust the code and merge it later. We are using this on real project, so I should be confindent it works as expected. Thank you once again

**Comment by nandordudas** [comment]
If you have any questions let me know and @jarcko please don't forget to remove me from collaborators asap/when possible.

**Comment by nandordudas** [comment]
Note: you can run benchmark locally with `pnpm run bench`.

**Review by jarcko — APPROVED** [review]

**Pull request merged**